File: go/init_test.go
/* Copyright © 2021-2026 Voxgig Ltd, MIT License. */

package model

import (
	"os"
	"path/filepath"
	"testing"
)

// Init scaffolds the starter files, and skips them on a second run.
func TestInitScaffolds(t *testing.T) {
	dir := t.TempDir()

	created, skipped, err := Init(dir)
	if err != nil {
		t.Fatal(err)
	}
	if len(created) != 2 || len(skipped) != 0 {
		t.Fatalf("created=%v skipped=%v", created, skipped)
	}
	for _, p := range []string{
		filepath.Join(dir, "model", "model.jsonic"),
		filepath.Join(dir, "model", ".model-config", "model-config.jsonic"),
	} {
		if _, err := os.Stat(p); err != nil {
			t.Fatalf("expected %s to exist: %v", p, err)
		}
	}

	created2, skipped2, err := Init(dir)
	if err != nil {
		t.Fatal(err)
	}
	if len(created2) != 0 || len(skipped2) != 2 {
		t.Fatalf("re-init created=%v skipped=%v (should skip existing)", created2, skipped2)
	}
}

// A freshly scaffolded project builds successfully.
func TestInitThenBuild(t *testing.T) {
	dir := t.TempDir()
	if _, _, err := Init(dir); err != nil {
		t.Fatal(err)
	}
	m := New(ModelSpec{
		Path: filepath.Join(dir, "model", "model.jsonic"),
		Base: filepath.Join(dir, "model"),
	})
	if br := m.Run(); !br.OK {
		t.Fatalf("scaffolded model failed to build: %v", br.Errs)
	}
	if m.Build().Model["name"] != "my-model" {
		t.Fatalf("scaffolded model = %#v", m.Build().Model)
	}
}

File: ts/test/init.test.ts
/* Copyright © 2021-2025 Voxgig Ltd, MIT License. */

import Fs from 'node:fs'
import { rm, stat } from 'node:fs/promises'
import { spawnSync } from 'node:child_process'
import { test, describe } from 'node:test'
import assert from 'node:assert'

import { Model, initModel } from '../dist/model'


const GEN = __dirname + '/../test/_gen'
const BIN = __dirname + '/../bin/voxgig-model'


describe('init', () => {

  test('scaffolds-and-skips-existing', async () => {
    const dir = GEN + '/init01'
    await rm(dir, { recursive: true, force: true })

    const r1 = initModel(dir, Fs)
    assert.strictEqual(r1.created.length, 2)
    assert.strictEqual(r1.skipped.length, 0)
    await stat(dir + '/model/model.jsonic')
    await stat(dir + '/model/.model-config/model-config.jsonic')

    // Second run leaves existing files untouched.
    const r2 = initModel(dir, Fs)
    assert.strictEqual(r2.created.length, 0)
    assert.strictEqual(r2.skipped.length, 2)
  })


  test('scaffold-builds', async () => {
    const dir = GEN + '/init02'
    await rm(dir, { recursive: true, force: true })
    initModel(dir, Fs)

    const model = new Model({
      path: dir + '/model/model.jsonic', base: dir + '/model', debug: 'silent',
    })
    const br = await model.run()
    assert.ok(br.ok, 'scaffolded model failed: ' + JSON.stringify(br.errs))
  })


  test('cli-init', async () => {
    const dir = GEN + '/init03'
    await rm(dir, { recursive: true, force: true })

    const res = spawnSync(process.execPath, [BIN, 'init', dir], { encoding: 'utf8' })
    assert.strictEqual(res.status, 0, res.stderr)
    assert.ok(res.stdout.includes('created:'), res.stdout)
    await stat(dir + '/model/model.jsonic')
  })

})
